testing/test_session.py

```diff
@@ -335,6 +335,54 @@ def pytest_sessionfinish():
     assert res.ret == ExitCode.NO_TESTS_COLLECTED
 
 
+def test_collection_args_do_not_duplicate_modules(pytester: Pytester) -> None:
+    """Test that when multiple collection args are specified on the command line
+    for the same module, only a single Module collector is created.
+
+    Regression test for #723, #3358.
+    """
+    pytester.makepyfile(
+        **{
+            "d/test_it": """
+                def test_1(): pass
+                def test_2(): pass
+                """
+        }
+    )
+
+    result = pytester.runpytest(
+        "--collect-only",
+        "d/test_it.py::test_1",
+        "d/test_it.py::test_2",
+    )
+    result.stdout.fnmatch_lines(
+        [
+            "<Module d/test_it.py>",
+            "  <Function test_1>",
+            "  <Function test_2>",
+        ],
+        consecutive=True,
+    )
+
+    # Different, but related case.
```

Comment:
I added this case as well, in relation to `node_cache2` in `Session.collect()`.

But I wonder if this is actually the behavior we want. If a user passes `d` twice with `--keep-duplicates`, do they expect e.g. module fixtures the be run once or twice?

Reply:
I would say run once makes more sense, I consider a "module" fixture being executed once to be an invariant; the keep duplicates only is there to "duplicate" the tests themselves, not fixture execution.

But I'm only really guessing.

Reply:
I feel similarly - it seems better but not quite sure. Anyway, I figure if we change something here it should be a conscious choice so the test is good to have.